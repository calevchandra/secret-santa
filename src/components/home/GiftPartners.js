import React, { useState } from "react";
import PropTypes from "prop-types";
import { Container, Button, Table, Alert } from "react-bootstrap";
import { toast } from "react-toastify";

import { API_BASE_URL } from "../../config";

const GiftPartners = ({
  assignGiftPartners,
  giftExchange,
  resetGiftExchange
}) => {
  const [disableAssignButton, setDisableAssignButton] = useState(false);

  /**
   * sendToAll sends an email to each of the participant,
   * informing who they are gifting to in the secret santa exchange
   */
  const sendToAll = () => {
    giftExchange.map(data => {
      sendEmailAlert(data.name, data.email, data.giftTo);
    });
  };

  return (
    <Container style={{ padding: 20 }}>
      <div>
        <Button
          disabled={disableAssignButton}
          onClick={() => {
            assignGiftPartners();
            setDisableAssignButton(true);
          }}
          variant="dark"
        >
          Assign
        </Button>
      </div>
      {giftExchange && giftExchange.length > 1 ? (
        <>
          <Table borderless responsive striped bordered hover size="sm">
            <thead>
              <tr>
                <th>Name</th>
                <th>Gifting to</th>
              </tr>
            </thead>
            <tbody>
              {giftExchange.map((data, idx) => {
                return (
                  <tr key={idx}>
                    <td>
                      {data.name.charAt(0).toUpperCase() + data.name.slice(1)}
                    </td>
                    <td>
                      {data.giftTo.charAt(0).toUpperCase() +
                        data.giftTo.slice(1)}
                    </td>
                  </tr>
                );
              })}
            </tbody>
          </Table>

          <Button
            block
            onClick={() => {
              sendToAll();
              setDisableAssignButton(false);
              resetGiftExchange();
            }}
            variant="info"
          >
            Notify
          </Button>
        </>
      ) : (
        <div style={{ paddingTop: 10 }}>
          <Alert variant="info">Staffs not allocated to a partner</Alert>
        </div>
      )}
    </Container>
  );
};

/**
 * sendEmailAlert calls an API that uses sendGrid to send emails.
 * The API server automatically starts with 'npm start'
 */
const sendEmailAlert = (name, email, giftingTo) => {
  return fetch(`${API_BASE_URL}send-email/`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      "content-type": "application/json"
    },
    body: JSON.stringify({
      recipient: email,
      sender: "cbkavit@gmail.com",
      topic: "Secret Santa",
      text: `Hi ${name.charAt(0).toUpperCase() +
        name.slice(1)}, you are selected to gift ${giftingTo
        .charAt(0)
        .toUpperCase() + giftingTo.slice(1)}. All the best `
    })
  })
    .then(response => response.json())
    .then(data => {
      if (data.msg === "Sent") toast.success("Email sent.");
    })
    .catch(error => error.message);
};

GiftPartners.propTypes = {
  assignGiftPartners: PropTypes.func.isRequired,
  giftExchange: PropTypes.array.isRequired,
  resetGiftExchange: PropTypes.func.isRequired
};

export default GiftPartners;
