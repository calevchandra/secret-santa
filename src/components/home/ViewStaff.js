import React, { useState } from "react";
import PropTypes from "prop-types";
import { Button, Table, FormControl } from "react-bootstrap";

const ViewStaff = ({
  handleEditChange,
  deleteParticipant,
  saveParticipant,
  staffData,
  formValid
}) => {
  const [indexValue, setIndexValue] = useState("");
  const [edit, setEdit] = useState(false);

  return (
    <Table striped bordered hover>
      <thead>
        <tr>
          <th>Name</th>
          <th>Email</th>
          <th>Action</th>
        </tr>
      </thead>
      <tbody>
        {staffData
          ? staffData.map((data, idx) => {
              return (
                <tr key={idx}>
                  <td>
                    {edit && indexValue === idx ? (
                      <FormControl
                        className={
                          !formValid.staffName ? "border border-danger" : null
                        }
                        name="editStaffName"
                        value={data.name}
                        onChange={e => handleEditChange(data.id, e)}
                      />
                    ) : (
                      data.name.charAt(0).toUpperCase() + data.name.slice(1)
                    )}
                  </td>
                  <td>
                    {edit && indexValue === idx ? (
                      <FormControl
                        className={
                          !formValid.staffEmail ? "border border-danger" : null
                        }
                        name="editStaffEmail"
                        value={data.email}
                        onChange={e => handleEditChange(data.id, e)}
                      />
                    ) : (
                      data.email
                    )}
                  </td>
                  <td>
                    {edit && indexValue === idx ? (
                      <Button
                        onClick={() => {
                          saveParticipant(data.name, data.email);
                          formValid.staffEmail ? setEdit(false) : setEdit(true);
                          setIndexValue("");
                        }}
                        variant="success"
                      >
                        Save
                      </Button>
                    ) : (
                      <>
                        <Button
                          onClick={() => {
                            setEdit(true);
                            setIndexValue(idx);
                            //enableEditParticipant(data.name, data.email)
                          }}
                          style={{ paddingRight: 20 }}
                          variant="warning"
                        >
                          Edit
                        </Button>
                        <Button
                          onClick={() => deleteParticipant(idx)}
                          variant="info"
                        >
                          Delete
                        </Button>
                      </>
                    )}
                  </td>
                </tr>
              );
            })
          : null}
      </tbody>
    </Table>
  );
};

ViewStaff.propTypes = {
  handleEditChange: PropTypes.func.isRequired,
  deleteParticipant: PropTypes.func.isRequired,
  saveParticipant: PropTypes.func.isRequired,
  staffData: PropTypes.array.isRequired,
  formValid: PropTypes.oneOfType([PropTypes.array, PropTypes.object])
};

export default ViewStaff;
