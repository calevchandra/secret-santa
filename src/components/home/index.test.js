import React from "react";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import { configure, shallow } from "enzyme";
import ViewStaff from "./ViewStaff";
import GiftPartners from "./GiftPartners";
import Home from "./index";

configure({ adapter: new Adapter() });

const test_staff_data = [
  { id: "_uhkqzgaos", name: "calev", email: "cbkavit@gmail.com" },
  { id: "_fcjnf1rkv", name: "caleb", email: "cbkavit@gmail.com" },
  { id: "_1aku7ekkn", name: "john", email: "cbkavit@gmail.com" },
  { id: "_gdiaqc0l6", name: "jerry", email: "cbkavit@gmail.com" },
  { id: "_qvl7ofesg", name: "sera", email: "cbkavit@gmail.com" },
  { id: "_v166hbko0", name: "tui", email: "cbkavit@gmail.com" },
  { id: "_zkesxrlpw", name: "jane", email: "cbkavit@gmail.com" },
  { id: "_x49fblif7", name: "jack", email: "cbkavit@gmail.com" },
  { id: "_7ugu9xzpj", name: "sam", email: "cbkavit@gmail.com" },
  { id: "_li6iqjfe7", name: "bob", email: "cbkavit@gmail.com" }
];

let home = new Home();

const minPropsViewStaffs = {
  deleteParticipant: () => {},
  handleEditChange: () => {},
  saveParticipant: () => {},
  staffData: test_staff_data,
  formValid: []
};

const minPropsGiftPartners = {
  assignGiftPartners: () => {},
  giftExchange: [],
  resetGiftExchange: () => {}
};

describe("Main Page", () => {
  it("renders without exploding", () => {
    expect(shallow(<Home />).length).toEqual(1);
  });
  it("renders as expected", () => {
    const tree = renderer.create(<Home />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe("The View Staff list", () => {
  it("renders without exploding", () => {
    expect(shallow(<ViewStaff {...minPropsViewStaffs} />).length).toEqual(1);
  });
  it("renders as expected", () => {
    const tree = renderer
      .create(<ViewStaff {...minPropsViewStaffs} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
  /*it("should add staff to the listing", () => {
    //expect(true).toEqual(true);
    // home.state.staff_data = test_staff_data;
    home.state({
      staff_data: test_staff_data,
      staffName: "Tim",
      emailAdd: "cbkavit@gmail.com"
    });
    home.handleAddParticipant({ target: { value: "h" } });
    expect(home.state.staff_data.length).toEqual(11);
  });*/
});

describe("Assign Partner list", () => {
  it("renders without exploding", () => {
    expect(shallow(<GiftPartners {...minPropsGiftPartners} />).length).toEqual(
      1
    );
  });
  it("renders as expected", () => {
    const tree = renderer
      .create(<GiftPartners {...minPropsGiftPartners} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
