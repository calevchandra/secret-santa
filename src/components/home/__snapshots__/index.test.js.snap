// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assign Partner list renders as expected 1`] = `
<div
  className="container"
  style={
    Object {
      "padding": 20,
    }
  }
>
  <div>
    <button
      className="btn btn-dark"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Assign
    </button>
  </div>
  <div
    style={
      Object {
        "paddingTop": 10,
      }
    }
  >
    <div
      className="fade alert alert-info show"
      role="alert"
    >
      Staffs not allocated to a partner
    </div>
  </div>
</div>
`;

exports[`Main Page renders as expected 1`] = `
<div>
  <div
    className="container"
  >
    <h1>
      Secret Santa
    </h1>
    <nav
      className="nav nav-tabs"
      onKeyDown={[Function]}
      role="tablist"
    >
      <a
        aria-controls={null}
        aria-selected={true}
        className="nav-item nav-link active"
        data-rb-event-key="list"
        href="#"
        id={null}
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
      >
        Participation
      </a>
      <a
        aria-controls={null}
        aria-disabled={true}
        aria-selected={false}
        className="nav-item nav-link disabled"
        data-rb-event-key="draw"
        href="#"
        id={null}
        onClick={[Function]}
        onKeyDown={[Function]}
        role="tab"
        tabIndex={-1}
      >
        Gift Partner
      </a>
    </nav>
    <div
      className="tab-content"
    >
      <div
        aria-hidden={false}
        aria-labelledby={null}
        className="tab-pane active"
        id={null}
        role="tabpanel"
        style={
          Object {
            "padding": 30,
          }
        }
      >
        <div
          className="card bg-secondary text-white"
        >
          <div
            className="card-body"
          >
            <form
              className=""
              style={
                Object {
                  "padding": 20,
                }
              }
            >
              <div
                className="form-row"
              >
                <div
                  className="form-group col"
                >
                  <input
                    className="form-control form-control-lg"
                    id="formGridName"
                    name="staffName"
                    onChange={[Function]}
                    placeholder="Enter Name"
                    type="input"
                    value=""
                  />
                </div>
                <div
                  className="form-group col"
                >
                  <input
                    className="form-control form-control-lg"
                    id="formGridEmail"
                    name="emailAdd"
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value=""
                  />
                </div>
              </div>
              <button
                className="btn btn-dark btn-block btn-lg"
                disabled={false}
                onClick={[Function]}
                type="submit"
              >
                Add
              </button>
            </form>
          </div>
        </div>
        <div
          style={
            Object {
              "paddingTop": 50,
            }
          }
        />
        <div
          className="fade alert alert-info show"
          role="alert"
        >
          No staffs added yet
        </div>
      </div>
      <div
        aria-hidden={true}
        aria-labelledby={null}
        className="tab-pane"
        id={null}
        role="tabpanel"
      >
        <div
          className="container"
          style={
            Object {
              "padding": 20,
            }
          }
        >
          <div>
            <button
              className="btn btn-dark"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Assign
            </button>
          </div>
          <div
            style={
              Object {
                "paddingTop": 10,
              }
            }
          >
            <div
              className="fade alert alert-info show"
              role="alert"
            >
              Staffs not allocated to a partner
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`The View Staff list renders as expected 1`] = `
<table
  className="table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Email
      </th>
      <th>
        Action
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Calev
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Caleb
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        John
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Jerry
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Sera
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Tui
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Jane
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Jack
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Sam
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
    <tr>
      <td>
        Bob
      </td>
      <td>
        cbkavit@gmail.com
      </td>
      <td>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "paddingRight": 20,
            }
          }
          type="button"
        >
          Edit
        </button>
        <button
          className="btn btn-info"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
